### PR TITLE
feat(bazel): Bazel workspace schematics

### PR DESCRIPTION
**Pull request by kyliau**
This commit creates a schematics for Bazel workspace.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [x] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,7 +1,9 @@
 .DS_STORE
 
 /dist/
-bazel-*
+bazel-bin
```
Comment (alexeagle) [inline .gitignore]: make sure this covers all the files

File: packages/bazel/src/schematics/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+filegroup(
+    name = "package_assets",
+    srcs = glob(["*"]),
+    visibility = ["//packages/bazel:__subpackages__"],
+)
+
+load("//tools:defaults.bzl", "ts_library", "jasmine_node_test")
+
+ts_library(
```
Comment (alexeagle) [inline packages/bazel/src/schematics/BUILD.bazel]: maybe just list them in the npm_package rule?

File: packages/bazel/src/schematics/README.md [hunk 3]
```diff
@@ -0,0 +1,12 @@
+## Generate .d.ts file from JSON schema
+
+The script to generate .d.ts file is located in the angular-cli repo. Make sure
+the CLI repository is checked out on your local machine.
+
+Then, in the CLI repository, run the following command
```
Comment (alexeagle) [inline packages/bazel/src/schematics/README.md]: should have a TODO to make the `ts_json_schema` rule re-usable and portable

File: packages/bazel/src/schematics/bazel-workspace/index.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+import { validateProjectName } from "@schematics/angular/utility/validation";
+import { Schema as BazelWorkspaceOptions } from './schema';
+import { strings } from '@angular-devkit/core';
+
+export default function (options: BazelWorkspaceOptions): Rule {
+  return (host: Tree, context: SchematicContext) => {
+    if (!options.name) {
+      throw new SchematicsException(`Invalid options, "name" is required.`);
+    }
+    validateProjectName(options.name);
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/index.ts]: does this match bazel's rules for workspace names? I imagine there's at least some small difference

File: packages/bazel/src/schematics/bazel-workspace/index.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      newProjectRoot = workspace.newProjectRoot || '';
+    } catch { }
+    const appDir = `${newProjectRoot}/${options.name}`;
+
+    const workspaceVersions = {
+      'ANGULAR_VERSION': '7.0.2',
+      'BAZEL_BUILDTOOLS_VERSION': '0.17.1',
+      'RULES_SASS_VERSION': '1.14.1',
+      'RXJS_VERSION': '6.3.3',
+      'NODE_VERSION': '10.9.0',
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/index.ts]: these should use the defaults

File: packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+load("@build_bazel_rules_typescript//:package.bzl", "rules_typescript_dependencies")
+rules_typescript_dependencies()
+# build_bazel_rules_nodejs is loaded transitively through rules_typescript_dependencies.
+
+load("@build_bazel_rules_nodejs//:defs.bzl", "check_bazel_version", "node_repositories", "yarn_install")
+# 0.18.0 is needed for .bazelignore
+check_bazel_version("0.18.0")
+node_repositories(
+    preserve_symlinks = True,
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template]: omit these three lines

File: packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template [hunk 7]
```diff
@@ -0,0 +1,88 @@
+# The WORKSPACE file tells Bazel that this directory is a "workspace", which is like a project root.
+# The content of this file specifies all the external dependencies Bazel needs to perform a build.
+
+####################################
+# ESModule imports (and TypeScript imports) can be absolute starting with the workspace name.
+# The name of the workspace should match the npm package where we publish, so that these
+# imports also make sense when referencing the published package.
+workspace(name = "<%= name %>")
+
+# The Bazel buildtools repo contains tools like the BUILD file formatter, buildifier
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template]: I think remove this one?

File: packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template [hunk 8]
```diff
@@ -0,0 +1,88 @@
+# The WORKSPACE file tells Bazel that this directory is a "workspace", which is like a project root.
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/WORKSPACE.template]: Needs a comment about how this is generated code and not meant to be edited unless you've learned Bazel

File: packages/bazel/src/schematics/bazel-workspace/index.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    validateProjectName(options.name);
+    let newProjectRoot = '';
+    try {
+      const workspace = getWorkspace(host);
+      newProjectRoot = workspace.newProjectRoot || '';
+    } catch { }
+    const appDir = `${newProjectRoot}/${options.name}`;
+
+    const workspaceVersions = {
+      'ANGULAR_VERSION': '7.0.2',
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/index.ts]: can this match the angular version from the package.json?

File: packages/bazel/src/schematics/bazel-workspace/index.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+      'BAZEL_BUILDTOOLS_VERSION': '0.17.1',
+      'RULES_SASS_VERSION': '1.14.1',
+      'RXJS_VERSION': '6.3.3',
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/index.ts]: would be nice to make it match what's in node_modules

File: packages/bazel/src/schematics/bazel-workspace/schema.d.ts [hunk 11]
```diff
@@ -0,0 +1,21 @@
+
+// THIS FILE IS AUTOMATICALLY GENERATED. TO UPDATE THIS FILE YOU NEED TO CHANGE THE
+// CORRESPONDING JSON SCHEMA FILE, THEN RUN devkit-admin build (or bazel build ...).
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/schema.d.ts]: comment doesn't make sense in this repo

File: packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# Make TypeScript and Angular compilation fast, by keeping a few copies of the
+# compiler running as daemons, and cache SourceFile AST's to reduce parse time.
+build --strategy=TypeScriptCompile=worker
+build --strategy=AngularTemplateCompile=worker
+
+# Don't create bazel-* symlinks in the WORKSPACE directory.
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template]: the comment here is inconsistent
I think the value `=dist/` is okay so just update the comment

File: packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+# These require .gitignore and may scare users.
+# Also, it's a workaround for https://github.com/bazelbuild/rules_typescript/issues/12
+# which affects the common case of having `tsconfig.json` in the WORKSPACE directory.
+#
+# Instead, you should run `bazel info bazel-bin` to find out where the outputs went.
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template]: "the output will appear in `dist/bin`
"you'll need to ignore the `bazel-out` folder that's created in the workspace root"

File: packages/bazel/src/schematics/bazel-workspace/files/BUILD.bazel.template [hunk 14]
```diff
@@ -0,0 +1,12 @@
+package(default_visibility = ["//visibility:public"])
+
+alias(
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/BUILD.bazel.template]: remove this rule

File: packages/bazel/src/schematics/bazel-workspace/files/src/BUILD.bazel.template [hunk 15]
```diff
@@ -0,0 +1,90 @@
+package(default_visibility = ["//visibility:public"])
+
+load("@angular//:index.bzl", "ng_module")
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_library", "ts_web_test_suite")
+load("@build_bazel_rules_nodejs//:defs.bzl", "rollup_bundle", "history_server")
+load("@build_bazel_rules_typescript//:defs.bzl", "ts_devserver")
+
+ng_module(
+    name = "src",
+    srcs = glob(["**/*.ts"], exclude = ["**/*.spec.ts", "test.ts"]),
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/src/BUILD.bazel.template]: remove test.ts

File: packages/bazel/src/schematics/bazel-workspace/files/src/BUILD.bazel.template [hunk 16]
```diff
@@ -0,0 +1,90 @@
+package(default_visibility = ["//visibility:public"])
```
Comment (alexeagle) [inline packages/bazel/src/schematics/bazel-workspace/files/src/BUILD.bazel.template]: generally we should leave lots of helpful comments in these files
and especially links to relevant documentation to help people bootstrap knowledge
doesn't need to be in this PR

OR
a comment saying "this file is hard to understand so don't bother"

**Review by alexeagle — APPROVED** [review]

File: packages/bazel/src/schematics/bazel-workspace/index.ts (hunk 5 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/index.ts]: removed NODE_VERSION and YARN_VERSION

File: .gitignore (hunk 1 above)
Comment (kyliau) [reply .gitignore]: changed to `/bazel-*`

File: packages/bazel/src/schematics/BUILD.bazel (hunk 2 above)
Comment (kyliau) [reply packages/bazel/src/schematics/BUILD.bazel]: done

File: packages/bazel/src/schematics/bazel-workspace/index.ts (hunk 4 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/index.ts]: Made sure workspace name is underscore-d.

File: packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template (hunk 12 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template]: updated comment

File: packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template (hunk 13 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/files/__dot__bazelrc.template]: done

File: packages/bazel/src/schematics/bazel-workspace/index.ts (hunk 9 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/index.ts]: there's quite a bit of work involved in this, I'll address this in the next PR.
Created https://angular-team.atlassian.net/browse/TOOL-471

File: packages/bazel/src/schematics/bazel-workspace/index.ts (hunk 10 above)
Comment (kyliau) [reply packages/bazel/src/schematics/bazel-workspace/index.ts]: there's quite a bit of work involved in this, I'll address this in the next PR.
Created https://angular-team.atlassian.net/browse/TOOL-471

File: packages/bazel/src/schematics/collection.json [hunk 17]
```diff
@@ -0,0 +1,12 @@
+{
+    "name": "@angular/bazel",
+    "version": "0.1",
+    "schematics": {
+      "bazel-workspace": {
+        "factory": "./bazel-workspace",
+        "schema": "./bazel-workspace/schema.json",
+        "description": "Setup Bazel workspace",
+        "hidden": true
```
Comment (cyrilletuzi) [inline packages/bazel/src/schematics/collection.json]: @kyliau Why is this schematics hidden? Tools like [Angular schematics extension for VS Code](https://github.com/cyrilletuzi/vscode-angular-schematics) filter schematics marked as hidden. Is it the desired behavior as you don't want users to use this yet?

Comment (kyliau) [reply packages/bazel/src/schematics/collection.json]: yes, this schematics is not ready to be used pending changes in the CLI (CLI does not support working on a subtree). For now it's meant to be called by `ng new` in `@angular-bazel` only.

**Pull request closed without merging**